modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/EnterpriseGeoIpDownloader.java

```diff
@@ -0,0 +1,449 @@
+                String name = entry.getKey();
+                Metadata meta = entry.getValue();
+                if (databases.contains(name) == false) {
+                    logger.debug("Dropping [{}], databases was {}", name, databases);
+                    _state = _state.remove(name);
+                    deleteOldChunks(name, meta.lastChunk() + 1);
+                    droppedSomething = true;
+                }
+            }
+            state = _state;
+            updateTaskState();
+        }
+
+        if (addedSomething == false && droppedSomething == false) {
+            RuntimeException accumulator = null;
+            for (Map.Entry<String, DatabaseConfigurationMetadata> entry : geoIpMeta.getDatabases().entrySet()) {
+                final String id = entry.getKey();
+                DatabaseConfiguration database = entry.getValue().database();
+
+                final String accountId = database.maxmind().accountId();
+                try (HttpClient.PasswordAuthenticationHolder holder = credentialsBuilder.apply(accountId)) {
+                    if (holder == null) {
+                        logger.warn("No credentials found to download database [{}], skipping download...", id);
+                    } else {
+                        processDatabase(holder.get(), id, database);
+                    }
+                } catch (Exception e) {
+                    accumulator = ExceptionsHelper.useOrSuppress(accumulator, ExceptionsHelper.convertToRuntime(e));
+                }
+            }
+            if (accumulator != null) {
+                throw accumulator;
+            }
+        }
+    }
+
+    void processDatabase(PasswordAuthentication auth, String id, DatabaseConfiguration database) throws IOException {
+        final String name = database.name();
+        logger.debug("Processing database [{}] for configuration [{}]", name, id);
+
+        final String sha256Url = downloadUrl(name, "tar.gz.sha256");
+        final String tgzUrl = downloadUrl(name, "tar.gz");
+
+        final Pattern checksumPattern = Pattern.compile("(\\w{64})\\s\\s(.*)");
+        String result = new String(httpClient.getBytes(auth, sha256Url), StandardCharsets.UTF_8).trim(); // this throws if the auth is bad
+        var matcher = checksumPattern.matcher(result);
+        boolean match = matcher.matches();
+        if (match == false) {
+            throw new RuntimeException("Unexpected sha256 response from [" + sha256Url + "]");
+        }
+        final String sha256 = matcher.group(1);
+        // the name that comes from the enterprise downloader cluster state doesn't include the .mmdb extension,
+        // but the downloading and indexing of database code expects it to be there, so we add it on here before further processing
+        processDatabase(auth, name + ".mmdb", sha256, tgzUrl);
+    }
+
+    private void processDatabase(PasswordAuthentication auth, String name, String sha256, String url) {
```

Comment:
Same here about javadoc, since the two `processDatabase` calls do different things (it may be useful to name them different things)

Reply:
I added javadocs. I'll leave it to @joegallo if he wants to come up with meaningful distinct names for these two methods.

Reply:
I'd prefer we didn't rename them -- there's a single implementation peeking out from underneath both of them that didn't get DRYed out on this PR. It is my intention to do that DRY-ing on a subsequent PR. I'll give them unique names for this PR if you insist, but they'll still be DRYed into a single name (and implementation) within month from now anyway.

Reply:
Leaving it is fine with me, can you add javadoc to both (even though they're private), so someone can tell how they differ at a glance?

Reply:
Certainly!

Reply:
Keith added javadocs in 225513ab6d6 and I expanded them slightly in 3f3a7c0849e.